Mathlib/Combinatorics/CombinatorialMap.lean

```diff
@@ -0,0 +1,187 @@
+def opp (M : CombinatorialMap D) : CombinatorialMap D where
+  σ := M.α * M.σ * M.α
+  α := M.α
+  φ := M.α * M.φ * M.α
+  involutive := M.involutive
+  fixedPoints_isEmpty := M.fixedPoints_isEmpty
+  composition := by
+    ext x
+    simp only [Equiv.Perm.coe_mul, Function.comp_apply, Equiv.Perm.coe_one, id_eq]
+    rw [M.involutive, M.involutive, composition'_apply]
+
+lemma opp_equiv (M : CombinatorialMap D) : ∃ (f : Equiv D D), Hom M M.opp f :=
+  ⟨M.α, ⟨by rw [← (M.α ∘ M.σ).comp_id, ← Function.RightInverse.id M.involutive]; rfl, rfl,
+  by rw [← (M.α ∘ M.φ).comp_id, ← Function.RightInverse.id M.involutive]; rfl⟩⟩
+
+/-- The dual of a `CombinatorialMap` swaps the roles of vertices and faces -/
+def dual (M : CombinatorialMap D) : CombinatorialMap D where
+  σ := M.φ⁻¹
+  α := M.α⁻¹
+  φ := M.σ⁻¹
+  composition := by
+    ext x
+    apply_fun M.α * M.φ * M.σ
+    simp [composition'_apply]
+  involutive := by
+    have : M.α⁻¹ = M.α := M.involutive.symm_eq_self_of_involutive
+    simp [this, M.involutive]
+  fixedPoints_isEmpty := by
+    have : M.α⁻¹ = M.α := M.involutive.symm_eq_self_of_involutive
+    simp [this, M.fixedPoints_isEmpty]
+
+/-- The double dual of a `CombinatorialMap` is equal to the original map. -/
+lemma double_dual_eq (M : CombinatorialMap D) : M.dual.dual = M := rfl
+
+/-- The vertices of a `CombinatorialMap`. -/
+@[reducible] def vertices (M : CombinatorialMap D) :=
+  Quotient (Equiv.Perm.SameCycle.setoid M.σ)
+
+/-- The edges of a `CombinatorialMap`. -/
+@[reducible] def edges (M : CombinatorialMap D) :=
+  Quotient (Equiv.Perm.SameCycle.setoid M.α)
+
+/-- The faces of a `CombinatorialMap`. -/
+@[reducible] def faces (M : CombinatorialMap D) :=
+  Quotient (Equiv.Perm.SameCycle.setoid M.φ)
+
+/-- The vertex corresponding to a dart `d`. -/
+@[reducible] def dartVertex (M : CombinatorialMap D) (d : D) : M.vertices :=
+  Quotient.mk (Equiv.Perm.SameCycle.setoid M.σ) d
+
+/-- The edge corresponding to a dart `d`. -/
+@[reducible] def dartEdge (M : CombinatorialMap D) (d : D) : M.edges :=
+  Quotient.mk (Equiv.Perm.SameCycle.setoid M.α) d
+
+/-- The face corresponding to a dart `d`. -/
+@[reducible] def dartFace (M : CombinatorialMap D) (d : D) : M.faces :=
+  Quotient.mk (Equiv.Perm.SameCycle.setoid M.φ) d
+
+/-- The degree of a vertex is the number of incident darts. -/
+def vertices.deg [Fintype D] [DecidableEq D] (v : M.vertices) : ℕ :=
+  Quotient.lift (fun w ↦ Fintype.card {u | Equiv.Perm.SameCycle M.σ w u})
+    (by
+      intro w u h
+      simp only [Set.coe_setOf, Set.mem_setOf_eq]
+      convert_to (fun x ↦ M.σ.SameCycle w x) = fun x ↦ M.σ.SameCycle u x
+      · apply Iff.intro
+        · intro
+          ext
+          exact ⟨h.symm.trans, h.trans⟩
+        · simp_all
+      ext
+      exact ⟨h.symm.trans, h.trans⟩) v
+
+/-- Set of darts incident to a vertex. -/
+def vertices.darts (v : M.vertices) : Set D :=
+  {d : D | M.dartVertex d = v}
```

Comment:
Use `Set.preimage`.

Or, since nothing uses this definition yet, we can delete it and have users write `M.dartVertex ⁻¹' {v}` themselves.